tools/make_patch.c

```diff
@@ -0,0 +1,450 @@
+			if (value > 0xffff) {
+				error_exit("Error: Invalid value for '%s' argument: %d", command, value);
+			}
+			fprintf(output, isupper((unsigned char)command[0]) ? " %02X %02X": " %02x %02x", value & 0xff, value >> 8);
+		}
+
+	} else if (!strcmp(command, "db") || !strcmp(command, "DB")) {
+		if (argc != 1) {
+			error_exit("Error: Invalid arguments for command: %s", command);
+		}
+		int value = parse_arg_value(argv[0], false, symbols, current_hook->name);
+		if (value > 0xff) {
+			error_exit("Error: Invalid value for '%s' argument: %d", command, value);
+		}
+		fprintf(output, isupper((unsigned char)command[0]) ? "a1:%02X": "a1:%02x", value);
+
+	} else if (!strcmp(command, "hex") || !strcmp(command, "HEX") || !strcmp(command, "HEx") || !strcmp(command, "Hex") || !strcmp(command, "heX") || !strcmp(command, "hEX")) {
+		if (argc != 1 && argc != 2) {
+			error_exit("Error: Invalid arguments for command: %s", command);
+		}
+		int value = parse_arg_value(argv[0], true, symbols, current_hook->name);
+		int padding = argc > 1 ? parse_number(argv[1], 0) : 2;
+		if (!strcmp(command, "HEx")) {
+			fprintf(output, "0x%0*X%02x", padding - 2, value >> 8, value & 0xff);
+		} else if (!strcmp(command, "Hex")) {
+			fprintf(output, "0x%0*X%03x", padding - 3, value >> 12, value & 0xfff);
+		} else if (!strcmp(command, "heX")) {
+			fprintf(output, "0x%0*x%02X", padding - 2, value >> 8, value & 0xff);
+		} else if (!strcmp(command, "hEX")) {
+			fprintf(output, "0x%0*x%03X", padding - 3, value >> 12, value & 0xfff);
+		} else {
+			fprintf(output, isupper((unsigned char)command[0]) ? "0x%0*X" : "0x%0*x", padding, value);
+		}
+
+	} else {
+		error_exit("Error: Unknown command: %s\n", command);
+	}
+}
+
+struct Buffer *process_template(const char *template_filename, const char *patch_filename, FILE *new_rom, FILE *orig_rom, const struct Symbol *symbols) {
+	FILE *input = xfopen(template_filename, 'r');
+	FILE *output = xfopen(patch_filename, 'w');
+	struct Buffer *buffer = buffer_create(1);
+
+	struct Buffer *patches = buffer_create(sizeof(struct Patch));
+
+	// The ROM checksum will always differ
+	buffer_append(patches, &(struct Patch){0x14e, 2});
+	// The Stadium data (see stadium.c) will always differ
+	unsigned int rom_size = (unsigned int)xfsize("", orig_rom);
+	unsigned int stadium_size = 24 + 6 + 2 + (rom_size / 0x2000) * 2;
+	buffer_append(patches, &(struct Patch){rom_size - stadium_size, stadium_size});
+
+	const struct Symbol *current_hook = NULL;
+
+	// Fill in the template
+	for (int c = getc(input); c != EOF; c = getc(input)) {
+		switch (c) {
+		case ';':
+			// ";" comments until the end of the line
+			putc(c, output);
+			for (c = getc(input); c != EOF; c = getc(input)) {
+				putc(c, output);
+				if (c == '\n' || c == '\r') {
+					break;
+				}
+			}
+			break;
+
+		case '{':
+			// "{...}" is a template command; buffer its contents
+			buffer->size = 0;
+			for (c = getc(input); c != EOF && c != '}'; c = getc(input)) {
```

Comment:
Getting EOF should probably not terminate this loop (or the other ones) cleanly

Reply:
Meh, if a file ends with `[Foo` or `{db 42` I'm okay with it acting like the correct closing delimited was there.

Reply:
Since EOL is not accounted for, it's more like a brace/bracket was left unclosed on accident.

Reply:
EOL is not generally significant. It's only significant for ending comments, and then it definitely is okay to omit the closing `\r` or `\n`, since a file might end in `; some comment without a newline`.

Template `{commands}` can include newlines, which some people might want for spacing out many `dws` with one per line. Patch `[names]` can too, which is less useful, but it's not like we're filtering out any other invalid patch name characters; `[!@#$%^&*()_+]` is parsed just fine.